File: 0644-maximum-average-subarray-ii/0644-maximum-average-subarray-ii.py
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        '''
        res = float('-inf')
        prefix = [0]
        for i in range(len(nums)):
            prefix.append(prefix[-1] + nums[i])
        
        for i in range(k, len(nums)+1):
            l = 0
            r = i
            while r < len(prefix):
                res = max(res, (prefix[r] - prefix[l]) / i)
                l+=1
                r+=1
        return res
        ''' 
            
            
        
        
        def isGood(val):
            currSum = 0
            prevSum = 0
            minSum = 0
            
            for i in range(len(nums)):
                currSum += nums[i] - val
                
                if i >= k-1 and currSum - minSum >= 0:
                    return True
                elif i >= k-1:
                    prevSum += nums[i-k+1] - val
                    minSum = min(minSum,prevSum)
                    
            
            return False
                    
        
        l,r = min(nums), max(nums)
        while r - l > 1e-5:
            
            mid = (l+r) / 2
            
            if isGood(mid):
                l = mid
            
            else:
                r = mid
        
        
        return l
        

File: 0644-maximum-average-subarray-ii/0644-maximum-average-subarray-ii.rs
impl Solution {
    pub fn find_max_average(nums: Vec<i32>, k: i32) -> f64 {
        let mut res = f64::MIN;
        let mut prefix = vec![0];
        for &num in &nums {
            prefix.push(*prefix.last().unwrap() + num);
        }

        for i in k as usize..=nums.len() {
            let mut l = 0;
            let mut r = i;
            while r < prefix.len() {
                res = res.max((prefix[r] as f64 - prefix[l] as f64) / i as f64);

                l += 1;
                r += 1;
            }
        }
        res
}

    
}